### PR TITLE
Process string functions XXXUTF8 'asciily' if input strings are all ascii chars

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Process string functions XXXUTF8 'asciily' if input strings are all ascii chars.  Inspired by https://github.com/apache/doris/pull/29799. Overall speed up by 1.07x~1.62x. Notice that peak memory usage had been decreased in some cases. 

``` sql 
Q1(1.27x for substringUTF8): 
select substringUTF8(materialize('hello world'), 2, 5) from numbers(100000000) format Null;
Before:
0 rows in set. Elapsed: 3.052 sec. Processed 100.00 million rows, 800.00 MB (32.77 million rows/s., 262.16 MB/s.)
Peak memory usage: 4.76 MiB.
After:
0 rows in set. Elapsed: 2.409 sec. Processed 100.00 million rows, 800.00 MB (41.52 million rows/s., 332.13 MB/s.)
Peak memory usage: 4.26 MiB.



Q2(1.24x for substringIndexUTF8): 
select substringIndexUTF8(materialize('www.clickhouse.com'), '.', 2) from numbers(100000000) format Null;
Before:
0 rows in set. Elapsed: 4.784 sec. Processed 100.00 million rows, 800.00 MB (20.90 million rows/s., 167.23 MB/s.)
Peak memory usage: 9.76 MiB.
After:
0 rows in set. Elapsed: 3.861 sec. Processed 100.00 million rows, 800.00 MB (25.90 million rows/s., 207.21 MB/s.)
Peak memory usage: 5.26 MiB. 


Q3(1.62x for reverseUTF8): 
select reverseUTF8(materialize('hello world')) from numbers(100000000) format Null; 
Before: 
0 rows in set. Elapsed: 3.044 sec. Processed 100.00 million rows, 800.00 MB (32.85 million rows/s., 262.80 MB/s.)
Peak memory usage: 4.51 MiB.
After: 
0 rows in set. Elapsed: 1.882 sec. Processed 100.00 million rows, 800.00 MB (53.14 million rows/s., 425.10 MB/s.)
Peak memory usage: 4.51 MiB. 


Q4(1.07x for lowerUTF8):
select lowerUTF8(materialize('hello world')) from numbers(100000000) format Null; 
Before: 
0 rows in set. Elapsed: 1.128 sec. Processed 100.00 million rows, 800.00 MB (88.64 million rows/s., 709.12 MB/s.)
Peak memory usage: 4.51 MiB.
After: 
0 rows in set. Elapsed: 1.073 sec. Processed 100.00 million rows, 800.00 MB (93.19 million rows/s., 745.55 MB/s.)
Peak memory usage: 4.50 MiB.


Q5(1.07x for upperUTF8):
select upperUTF8(materialize('hello world')) from numbers(100000000) format Null;
Before: 
0 rows in set. Elapsed: 1.137 sec. Processed 100.00 million rows, 800.00 MB (87.92 million rows/s., 703.37 MB/s.)
Peak memory usage: 4.51 MiB. 
After: 
0 rows in set. Elapsed: 1.067 sec. Processed 100.00 million rows, 800.00 MB (93.71 million rows/s., 749.65 MB/s.)
Peak memory usage: 4.50 MiB. 


Q6(1.23x for leftPadUTF8): 
select leftPadUTF8(materialize('hello '), 10, ',') from numbers(100000000) format Null; 
Before: 
0 rows in set. Elapsed: 4.322 sec. Processed 100.00 million rows, 800.00 MB (23.14 million rows/s., 185.11 MB/s.)
Peak memory usage: 5.01 MiB.
0 rows in set. Elapsed: 3.503 sec. Processed 100.00 million rows, 800.00 MB (28.55 million rows/s., 228.37 MB/s.)
Peak memory usage: 4.65 MiB.


Q7(1.28x for rightPadUTF8): 
Before:
select rightPadUTF8(materialize('hello '), 10, ',') from numbers(100000000) format Null; 
0 rows in set. Elapsed: 4.522 sec. Processed 100.00 million rows, 800.00 MB (22.11 million rows/s., 176.90 MB/s.)
After:
Peak memory usage: 5.01 MiB.
0 rows in set. Elapsed: 3.539 sec. Processed 100.00 million rows, 800.00 MB (28.26 million rows/s., 226.06 MB/s.)
Peak memory usage: 4.65 MiB.

```

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 49422debafbf773c887c60f2affb2e038edea911 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61632/49422debafbf773c887c60f2affb2e038edea911/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by evillique** [comment]
How does this optimization influence execution time if the string contains a non-ASCII character at the end of it, to check the worst-case scenario?

**Comment by taiyang-li** [comment]
> How does this optimization influence execution time if the string contains a non-ASCII character at the end of it, to check the worst-case scenario?

I understand your worry totally. But the time cost of `isAllASCII` is trivial. Let's add some cases of non-ASCII character to see how much they have been slowed down.

**Comment by taiyang-li** [comment]
@evillique the ci seems not stable. Could you review code changes first?

**Comment by taiyang-li** [comment]
@evillique Any comments?

File: src/Functions/padString.cpp [hunk 1]
```diff
@@ -236,23 +235,39 @@ namespace
...
+                    executeForSourceAndLength<false>(
+                        std::forward<SourceStrings>(strings), ConstSource<GenericValueSource>{*col_const}, padding_chars, res_sink);
+                else
+                    executeForSourceAndLength<false>(
+                        std::forward<SourceStrings>(strings), GenericValueSource{*column_length}, padding_chars, res_sink);
+            }
         }
 
-        template <typename SourceStrings, typename SourceLengths>
+        template <bool all_ascii, typename SourceStrings, typename SourceLengths>
```
Comment (al13n321) [inline src/Functions/padString.cpp:266]: Nitpick: seems weird that `all_ascii` has the same meaning as `is_utf8` but flipped. Maybe call it `is_actually_utf8` and remove the negation? (Or call it `is_utf8` and rename `is_utf8` to `support_utf8` or something.)

**Review by al13n321 — APPROVED** [review]
Looks reasonable, thanks for the improvement!

---

Wouldn't it be more efficient to do this as part of the function's inner loop instead? I.e. (1) load a block of characters into a simd register, (2) check if all ascii, (3a) use fast ascii path if yes, (3b) use slow utf8 path if no. The (1)+(2) can be identical to the `isAllASCII` implementation. The (3a) can be reused between the utf8 and the non-utf8 versions of the function. Overall it should be faster because it iterates only once and works for mixed strings.

Some of these functions already do this, but without AVX2/AVX512. Adding AVX2 and AVX512 implementations for these functions would be strictly better than the `isAllASCII` trick, but is more work, right? Specifically (from cursory reading, expect ~30% to be wrong):
 * LowerUpperUTF8Impl - just needs AVX2 and AVX512.
 * padString:
    - UTF8::countCodePoints needs AVX2 and AVX512
    - skipCodePointsBackward is not even simd
 * substring - skipCodePointsForward and skipCodePointsBackward need simd.
 * reverse - not obvious how to do better than `isAllASCII`.
 * substringIndex - I don't understand why this needs to be UTF8-aware at all; UTF8 is specifically designed such that regular memmem-like search works correctly (as long as needle is valid UTF8).

But the `isAllASCII` trick seems like a good default behavior for cases when when we don't have efficient utf8 implementation yet.

File: src/Functions/GatherUtils/Sinks.h [hunk 2]
```diff
@@ -108,7 +108,7 @@ struct StringSink
 
     void reserve(size_t num_elements)
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply src/Functions/GatherUtils/Sinks.h:109]: But I found ColumnVector::reserve invokes reserve_exact. It has the same issue as you describe above. Although I think executing both apis in a loop probably won't happen

File: src/Functions/padString.cpp (hunk 1 above)
Comment (taiyang-li) [reply src/Functions/padString.cpp:266]: all_ascii is true under two conditions:
- is_utf8 is false.
- is_utf8 is true, but its content is actually all ascii chars.

Comment (taiyang-li) [reply src/Functions/padString.cpp:266]: I'll rename all_ascii to is_actually_utf8

**Comment by al13n321** [comment]
TSAN error looks related: https://s3.amazonaws.com/clickhouse-test-reports/61632/690a8cf24b9402eac7479b0bbc62002b1ab3b7dc/stateless_tests__tsan__s3_storage__[4_5]/stderr.log . ASAN error seems to be the same: https://s3.amazonaws.com/clickhouse-test-reports/61632/690a8cf24b9402eac7479b0bbc62002b1ab3b7dc/stateless_tests__asan__[2_4]/stderr.log

I'm guessing it's the `memcpySmallAllowReadWriteOverflow15()`. When called from `constantVector()`, `res_ref` is not padded (points into `std::string`). Plz fix.

(The other 3 failed tests are flaky.)

**Pull request merged**